### DISCUSSION TITLE
Where is documentation for add_head_html and add_body_html?

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#2170] [Q&A]
### Question

This post contains an issue, an explanation, and suggested improvements.  These are low-effort short-term improvements that would make nicegui more mature and usable for web apps.  And likely cut down on forum posts asking the same things repeatedly.

# Issue
Where is the documentation for add_head_html and add_body_html?  They are recommended in several places in the docs (such as [here](https://nicegui.io/documentation/html) and [here](https://nicegui.io/documentation/section_text_elements) and [here](https://nicegui.io/documentation/generic_events)) and in the forums.  Yet I find no API, no call signatures, no explanation anywhere in the docs or forums.  Tried to generate my own docs but pydoc doesn't work on nicegui because of all the relative import statements:
```
problem in /Users/ed2050/lib/python/site-packages/nicegui/nicegui.py - 
ImportError: attempted relative import with no known parent package
```
Even running `help (ui.add_head_html)` in python console reveals almost nothing; just `(code: str) -> None`.   

# Why it Matters
I was recently bitten by an obscure add_head_html bug.  Trying to add a `<base href="http://somwhere.org">` tag to the header so relative links resolve correctly (displaying html from an external source).

For many attempts ui.add_head_html did absolutely nothing.  No error messages in terminal.  No messages in browser console.  No `<base>` tag in document tree in browser.  Call just silently did nothing.  I'm 100% certain the call to add_head_html was made correctly with the correct args, as confirmed by log messages immediately before and after.

I finally figured out add_head_html has to be called _before_ `await client.connected ()`.  Which is completely unexpected.  HTTP headers, sure, gotta send those before opening web sockets.  But HTML headers can be added to the dom at any time.  There's no reason nicegui can't add them after client.connected. 

If nicegui chooses not to do add_head_html for whatever reason, then _at least_ raise an error or warning in the console informing the user ("add_head_html: can't add headers after client.connected, see http://some.where/ for more info")

client.connected is required just to do simple things.  Like set a dynamic page title.  The entire content of your page can be added after client.connected () - ui.label, ui.tree, ui.fuzzywuzzy all work fine.  Only ui.add_head_html doesn't.

Actually that's not entirely true.  ui.add_body_html doesn't work either.  Whatever questionable justifications for not adding to HTML `<head>`, there's demonstrable evidence that adding to HTML `<body>` works perfectly fine after client.connected.  ui.image, ui.slider, ui.whozeit, no problems.  Even ui.html works great.  Unless you use ui.add_body_html.  Then things silently fail with no errors and no explanations.

# How to improve

## 1. Document basic interfaces
If add_head_html and add_body_html are recommended throughout the docs, they should be fully described in their own page including necessary caveats like "Doesn't work after client.connected".

## 2. Make nicegui compile correctly with pydocs.
So users can generate their own documentation where the website is lacking.  If this works with another standard doc tool, please mention that somewhere prominently.  Possibly on a page like "Need more help?" that points to the discussion  forum and other resources.

## 3. Support the minimum basic functionality for web apps
Any web app should have basic features to interact with the document.  These should be doable within nicegui, not require custom javascript snippets that are difficult to trace and debug.  This includes (not an exhaustive list) : 
- setting dynamic page titles.  If I have `@ui.page ('/foo/{name}')` then I may want `name` as part of my page title.
- giving focus to a certain DOM element
- scrolling a DOM element into view
- proper text elements: headers, paragraphs, etc.  nicegui is already moving in this direction from what I understand.

If nicegui is a toolkit for local apps that just happen to be built on HTML, then maybe Number 3 is unnecessary.  But nicegui seems to be being promoted as a suitable toolkit for building web apps.  Yet it lacks basic functionality every web developer needs.

I'm happy to help with these things.  I'm raising missing functionality when I come across it.  The forums are filled with more examples from users asking how to do certain things.

**Comment by retsyo** [comment]
I am too lazy to make a new post. I just write here, we can't search document on `ui.header`

**Reply by ed2050** [reply]
Hi retsyo, I don't follow you.  How do you "search" any document in nicegui?  Do you mean searching content in the DOM tree (the "document" in DOM parlance)?  Or searching nicegui objects?  How is ui.header different from other nicegui components in this regard?

**Reply by falkoschindler** [reply]
Event though you can search for "ui.header" and find the section on ["Page Layout"](https://nicegui.io/documentation/section_pages_routing#page_layout) mentioning it, a separate page with the complete API is indeed missing for these page layout elements.

**Reply by falkoschindler** [reply]
I finally added a reference page for page layout elements in https://github.com/zauberzeug/nicegui/commit/741f0a05cd77655d7eb6321a8718d6e568e62231
Once deployed, it will be located at https://nicegui.io/documentation/page_layout.

**Comment by stepheweffie** [comment]
```
def add_head_html(code: str) -> None:
    """Add HTML code to the head of the page."""
    context.get_client().head_html += code + '\n'
```
Runs on the slot stack. You have to read the code to understand what's going on here. 
<img width="1440" alt="Screen Shot 2023-12-27 at 4 27 35 PM" src="https://github.com/zauberzeug/nicegui/assets/1074503/cd288de6-4150-4445-ab50-fbfcd99dbfa1">

**Comment by falkoschindler** [comment]
I finally added a demo for `ui.add_head_html` in https://github.com/zauberzeug/nicegui/commit/ea18324b211e6243e771add3eba9f6bc8146a570.
Once deployed, it will be located at https://nicegui.io/documentation/section_pages_routing#add_html_to_the_page.
